core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala

```diff
@@ -234,19 +235,37 @@ object Protocol {
   }
 
   /**
-   * Extracts all table features that are enabled by the given metadata.
+   * Returns the smallest set of table features that contains `features` and that also contains
+   * all dependencies of all features in the returned set.
    */
-  def extractAutomaticallyEnabledFeatures(
-      spark: SparkSession, metadata: Metadata): Set[TableFeature] = {
-    val enabledFeatures = mutable.Set[TableFeature]()
-    TableFeature.allSupportedFeaturesMap.values.foreach {
-      case feature: TableFeature with FeatureAutomaticallyEnabledByMetadata =>
-        if (feature.metadataRequiresFeatureToBeEnabled(metadata, spark)) {
-          enabledFeatures += feature
-        }
-      case _ => () // Do nothing. We only care about features that can be activated in metadata.
+  @tailrec
+  private def getDependencyClosure(features: Set[TableFeature]): Set[TableFeature] = {
+    val requiredFeatures = features ++ features.flatMap(_.requiredFeatures)
+    if (features == requiredFeatures) {
+      features
+    } else {
+      getDependencyClosure(requiredFeatures)
     }
-    enabledFeatures.toSet
+  }
+
+  /**
+   * Extracts all table features that are enabled by the given protocol and metadata.
+   * This includes features enabled directly by metadata, and their dependencies.
+   */
+  def extractAutomaticallyEnabledFeatures(
+      spark: SparkSession, protocol: Protocol, metadata: Metadata): Set[TableFeature] = {
```

Comment:
IMO This method is doing too many things: it not only extracts automatically-enabled features, but also features from the input `protocol`, which is not always needed by the caller (for example, three call sites in this PR do not use this argument).

I suggest making this method do only one thing, by removing the  `protocol: Protocol` argument. We could make the `getDependencyClosure()` public if the caller wants to get dependencies from some existing protocol objects:

Instead of calling

```java
extractAutomaticallyEnabledFeatures(spark, protocol, metadata)
```

they can use 

```java
getDependencyClosure(protocol.implicitlyAndExplicitlySupportedFeatures) ++
  extractAutomaticallyEnabledFeatures(spark, metadata)
```
instead, which is more readable.

The `getDependencyClosure` method should be moved to `TableFeatureSupport.scala`, because it is a static helper method.

Reply:
It would have to be the following then:
```scala
TableFeatureSupport.getDependencyClosure(
  extractAutomaticallyEnabledFeatures(spark, metadata) ++ protocol.implicitlyAndExplicitlySupportedFeature
)
```

Reply:
I would prefer to keep this method as is. Otherwise it will be easy to forget to take the dependency closure.

Reply:
Done!